src/cli/util.js

```diff
@@ -397,33 +399,113 @@ function createIgnorerFromContextOrDie(context) {
   }
 }
 
-function eachFilename(context, patterns, callback) {
-  // The '!./' globs are due to https://github.com/prettier/prettier/issues/2110
-  const ignoreNodeModules = context.argv["with-node-modules"] !== true;
-  if (ignoreNodeModules) {
-    patterns = patterns.concat(["!**/node_modules/**", "!./node_modules/**"]);
+/**
+ * Get stats of a given path.
+ * @param {string} filePath The path to target file.
+ * @returns {fs.Stats|undefined} The stats.
+ * @private
+ */
+function statSafeSync(filePath) {
+  try {
+    return fs.statSync(filePath);
+  } catch (error) {
+    /* istanbul ignore next */
+    if (error.code !== "ENOENT") {
+      throw error;
+    }
   }
-  patterns = patterns.concat(["!**/.{git,svn,hg}/**", "!./.{git,svn,hg}/**"]);
+}
 
-  try {
-    const filePaths = globby
-      .sync(patterns, { dot: true, nodir: true })
-      .map(filePath => path.relative(process.cwd(), filePath));
+/**
+ * Check if a string is a glob pattern or not.
+ * @param {string} pattern A glob pattern.
+ * @returns {boolean} `true` if the string is a glob pattern.
+ */
+function isGlobPattern(pattern) {
+  return isGlob(path.sep === "\\" ? pattern.replace(/\\/gu, "/") : pattern);
+}
+
+const globbyOptions = { dot: true, nodir: true, absolute: true };
+function eachFilename(context, maybePatterns, callback) {
+  let files = [];
+  const cwd = process.cwd();
+  const patterns = [];
+
+  for (const pattern of maybePatterns) {
+    const absolutePath = path.resolve(cwd, pattern);
+    const stat = statSafeSync(absolutePath);
+
+    if (
+      stat &&
+      stat.isDirectory() &&
+      // `dot pattern` and `expand directories` support need handle differently
+      // for backward compatibility reason only expand `directories` like a glob pattern
+      // see https://github.com/prettier/prettier/pull/6639#issuecomment-548949954
+      isGlobPattern(pattern)
```

Comment:
Just infromation: `test*` can be directory and can be glob :smile: But maybe it is good idea to check on directory firstly, also we can use `https://github.com/mrmlnc/fast-glob#isdynamicpatternpattern-options` to check on glob, I think that's why I decided that we need to do an update first

Reply:
@evilebottnawi `is-glob` is original from `eslint`, we don't have `fast-glob` yet, we don't want both `fast-glob` and `globby`, we can switch to `globby` function later https://github.com/sindresorhus/globby#globbyhasmagicpatterns-options

Reply:
Yes, agree :+1:

Reply:
I don't understand this `if`. Why include `<dir>/**/*` if `<dir>` looks like a pattern, but is a directory? I read https://github.com/prettier/prettier/pull/6639#issuecomment-548949954, but it doesn't make it any clearer to me.

Reply:
Run 
```console
$ npx prettier src

[error] No matching files. Patterns tried: src !**/node_modules/** !./node_modules/** !**/.{git,svn,hg}/** !./.{git,svn,hg}/**
```

We don't expand dirname, unless it's pattern like `!dir` and it's a dir, we search all files in this `!dir`, because this is we want to do in this pr, otherwise, we still say we don't glob anything in `src`

Reply:
Let's say we have two directories (with formattable files inside): `foo` and `[foo]`.
Now, both `npx prettier foo` and `npx prettier [foo]` give the same result: "No matching files".

Why does this PR change the behavior for `prettier [foo]`? Sorry if it's something obvious, but somehow I can't see it.

Reply:
#6854 ask for support to find `[foo]` file, and @evilebottnawi want also support `[foo]` dir https://github.com/prettier/prettier/pull/6776#issuecomment-581363613

Reply:
If we don't expand this `[foo]` dir , people will still not able to format all files inside `[foo]` , because `prettier [foo]/**/*` will not glob any file, he has to `prettier [foo]/1.js [foo]/2.js [foo]/3.js`, but we do this, he can `prettier [foo]`. And we are going to support `prettier foo` #6128 , just not in this one.

Reply:
That's why there exists escaping syntax in globs. fast-glob uses backslashes for that, which, unfortunately, makes the Windows story messier... I'll check how ESLint handles this.

Reply:
What a mess. The same command `eslint "x\[foo\]/*.js"` matches different files on Linux (`x[foo] → foo.js`) and Windows (`x → [foo → ] → foo.js`).

Reply:
An [alternative escaping syntax](https://github.com/isaacs/minimatch/issues/64#issuecomment-258709901) works though: `eslint "x[[]foo]/*.js"`.

Reply:
> If we don't expand this [foo] dir , people will still not able to format all files inside [foo] ,

So, as we have the escaping syntax, this concern about 

> And we are going to support prettier foo #6128 , just not in this PR.

Let's move all the work with directories to that PR. For now, in this PR, let's just skip directories. To format files inside `[foo]`, people have the escaping syntax.